### PR TITLE
[AppService]: #7682- Added new cmdlets for managing Routing Rules

### PR DESCRIPTION
**Pull request by Kotasudhakarreddy**
## Description

WebApp Testing in Production Routing Rules not supported in RM Cmdlets #7682

## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
`Design Review is pending for approval.` https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/498
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by panchagnula** [comment]
@Kotasudhakarreddy you can squash all commits to a single one by running
git rebase -i HEAD~<number of your commits>
& force push
git push --force

Also in general you can rebase to upstream by
git pull --rebase upstream master 
this is better than merging upstream master to your branch

File: src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+using Microsoft.Azure.Management.WebSites.Models;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.WebApps.Cmdlets.TrafficRouting
+{
+    /// <summary>
+    /// this commandlet will let you Remove the given Azure App Service Traffic Routing using ARM APIs
+    /// </summary>
+    [Cmdlet(VerbsCommon.Remove, ResourceManager.Common.AzureRMConstants.AzurePrefix + "WebAppTrafficRouting", DefaultParameterSetName = RoutingParameterSet, SupportsShouldProcess = true), OutputType(typeof(RampUpRule))]
```
Comment (isra-fel) [inline src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs]:
```suggestion
    [Cmdlet(VerbsCommon.Remove, ResourceManager.Common.AzureRMConstants.AzurePrefix + "WebAppTrafficRouting", DefaultParameterSetName = RoutingParameterSet, SupportsShouldProcess = true), OutputType(typeof(bool))]
```
For a Remove cmdlet, the return type is bool

File: src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                    if (this.ShouldProcess(this.RuleName, string.Format("Deleting Routing Rule for slot - '{0}' from Web Application - {1}", this.RuleName, this.WebAppName)))
+                    {
+                        //Remove the rule
+                        siteConfig.Experiments.RampUpRules.Remove(givenRampUpRuleObj);
+                        // Update web app configuration
+                        WebsitesClient.UpdateWebAppConfiguration(ResourceGroupName, webApp.Location, WebAppName, null, siteConfig, null, null, null);
+                        if (PassThru.IsPresent)
+                        {
+                            var app = WebsitesClient.GetWebApp(ResourceGroupName, WebAppName, null);
+                            WriteObject(app.SiteConfig.Experiments.RampUpRules.FirstOrDefault(rule => rule.Name == RuleName));
```
Comment (isra-fel) [inline src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs]:
```suggestion
                            WriteObject(true);
```
And if passthru is present, just write `true`

File: src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs (hunk 1 above)
Comment (isra-fel) [reply src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs]: Please regenerate help after this change

File: src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                        }
+                    }
+                }
+                else
+                {
+                    throw new ValidationMetadataException(string.Format("Given Routing Rule with name '{0}' in WebApp '{1}' is not present." +
```
Comment (isra-fel) [inline src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs]: Please use resource strings for customer-facing messages.

File: src/Websites/Websites/Cmdlets/TrafficRouting/UpdateAzureWebAppTrafficRoutingRule.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                    {
+                        if (this.ShouldProcess(this.WebAppName, string.Format("Updating the Routing Rule for slot - '{0}' in Web Application - {1}.", rampUpRule.Name, this.WebAppName)))
+                        {
+                            //Remove previous rule
+                            siteConfig.Experiments.RampUpRules.Remove(givenRampUpRuleObj);
+                            //Add the updated rule
+                            siteConfig.Experiments.RampUpRules.Add(rampUpRule);
+                            // Update web app configuration
+                            WebsitesClient.UpdateWebAppConfiguration(ResourceGroupName, webApp.Location, WebAppName, null, siteConfig, null, null, null);
+                            //var app = WebsitesClient.GetWebApp(ResourceGroupName, WebAppName, null);
```
Comment (isra-fel) [inline src/Websites/Websites/Cmdlets/TrafficRouting/UpdateAzureWebAppTrafficRoutingRule.cs]: Please remove commented code that's not going to be used.

File: src/Websites/Websites/Cmdlets/TrafficRouting/UpdateAzureWebAppTrafficRoutingRule.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                            //WriteObject(app.SiteConfig.Experiments.RampUpRules.FirstOrDefault(rule => rule.Name == rampUpRule.Name));
+                            WriteObject(rampUpRule);
+                        }
+                    }
+                    else
+                    {
+                        throw new ValidationMetadataException(string.Format("A Routing Rule with name '{0}' in WebApp '{1}' already exists." +
```
Comment (isra-fel) [inline src/Websites/Websites/Cmdlets/TrafficRouting/UpdateAzureWebAppTrafficRoutingRule.cs]: Same thing about resource string

File: tools/StaticAnalysis/Exceptions/ValidateHelpIssues.csv [hunk 6]
```diff
@@ -1,140 +0,0 @@
-Target,Description
```
Comment (isra-fel) [inline tools/StaticAnalysis/Exceptions/ValidateHelpIssues.csv]: This file shouldn't be deleted.

File: src/Websites/Websites/Cmdlets/TrafficRouting/AddAzureWebAppTrafficRoutingRule.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                            // Update web app configuration
+                            WebsitesClient.UpdateWebAppConfiguration(ResourceGroupName, webApp.Location, WebAppName, null, siteConfig, null, null, null);
+                            //var app = WebsitesClient.GetWebApp(ResourceGroupName, WebAppName, null);
+                            //WriteObject(app.SiteConfig.Experiments.RampUpRules.FirstOrDefault(rule => rule.Name == rampUpRule.Name));
+                            WriteObject(rampUpRule);
+                        }
+                    }
+                    else
+                    {
+                        throw new ValidationMetadataException(string.Format("A Routing Rule with name '{0}' in WebApp '{1}' already exists." +
```
Comment (isra-fel) [inline src/Websites/Websites/Cmdlets/TrafficRouting/AddAzureWebAppTrafficRoutingRule.cs]: same thing about resource string

File: src/Websites/Websites/Cmdlets/TrafficRouting/GetAzureWebAppTrafficRoutingRule.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+                //Check for Rule existance with the given name.
+                var givenRampUpRuleObj = siteConfig.Experiments.RampUpRules.FirstOrDefault(item => item.Name == RuleName);
+                if (givenRampUpRuleObj != null)
+                {
+                    WriteObject(givenRampUpRuleObj);
+                }
+                else
+                {
+                    throw new ValidationMetadataException(string.Format("Given Routing Rule with name '{0}' in WebApp '{1}' is not present." +
```
Comment (isra-fel) [inline src/Websites/Websites/Cmdlets/TrafficRouting/GetAzureWebAppTrafficRoutingRule.cs]: resource string too

File: src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs (hunk 3 above)
Comment (isra-fel) [reply src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs]: It also avoids duplicated hardcoded string

Comment (Kotasudhakarreddy) [reply src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs]: > Please use resource strings for customer-facing messages.

Not clear on resource strings. Can you give me more details on it.

Comment (isra-fel) [reply src/Websites/Websites/Cmdlets/TrafficRouting/RemoveAzureWebAppTrafficRoutingRule.cs]: If you go to `Websites/Properties/Resoures.resx` in visual studio you'll find a dictionary of strings, simply add you string there and reference it by `Resources.{your-key}` in code.

Resource strings is a way to centralize string resources in your project. It's helpful for reducing code duplication and for i18n as well.

**Comment by isra-fel** [comment]
test

**Pull request merged**